src/libAtomVM/nifs.c

```diff
@@ -5119,6 +5125,99 @@ static term nif_unicode_characters_to_binary(Context *ctx, int argc, term argv[]
     return result_tuple;
 }
 
+static term nif_erlang_lists_subtract(Context *ctx, int argc, term argv[])
+{
+    UNUSED(argc)
+
+    term list1 = argv[0];
+    term list2 = argv[1];
+
+    if (!term_is_list(list1) || !term_is_list(list2)) {
+        return BADARG_ATOM;
+    }
+
+    int proper;
+    int len = term_list_length(list1, &proper);
+    if (UNLIKELY(!proper)) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    int proper2;
+    term_list_length(list2, &proper2);
+    if (UNLIKELY(!proper2)) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    if (term_is_nil(list1)) {
+        return term_nil();
+    }
+
+    if (term_is_nil(list2)) {
+        return list1;
+    }
+
+    term *cons = malloc(len * sizeof(term));
+    if (UNLIKELY(IS_NULL_PTR(cons))) {
+        RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+    }
+    int i = 0;
+    term list = list1;
+
+    while (!term_is_nil(list)) {
+        cons[i] = list;
+        list = term_get_list_tail(list);
+        i++;
+    }
+
+    int last_filtered_idx = -1;
+    term result = term_nil();
+
+    while (!term_is_nil(list2)) {
+        term to_nullify = term_get_list_head(list2);
+
+        for (int i = 0; i < len; i++) {
+            if (term_is_invalid_term(cons[i])) {
+                continue;
+            }
+            term item = term_get_list_head(cons[i]);
+            TermCompareResult cmp_result = term_compare(to_nullify, item, TermCompareExact, ctx->global);
+
+            if (UNLIKELY(cmp_result == TermCompareMemoryAllocFail)) {
+                free(cons);
+                RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+            }
+            if (cmp_result == TermEquals) {
+                if (last_filtered_idx < i) {
+                    last_filtered_idx = i;
+                    result = term_get_list_tail(cons[i]);
+                }
+                cons[i] = term_invalid_term();
+                break;
+            }
+        }
+        list2 = term_get_list_tail(list2);
+    }
+
+    if (last_filtered_idx == -1) {
+        free(cons);
+        return list1;
+    }
+
+    if (UNLIKELY(memory_ensure_free_with_roots(ctx, (last_filtered_idx + 1) * CONS_SIZE, 1, &list1, MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
```

Comment:
we need to use `cons` as roots, otherwise it will be invalid after a gc.

Reply:
Not sure this is true if list1 is in the roots.

Reply:
@pguyot of course every item in `cons` is also in `list1`, but we need array items in cons pointing to updated memory locations, otherwise we end having an array populated with invalid data after GC.

`cons` is required to build the `result` list  so we need `cons` to be valid, while `list1` is not required anymore just before gc.

Reply:
Also there is a problem with result, for which we are doing:
`result = term_get_list_tail(cons[i])`. I suggest storing the index we need to use somewhere else, and to declare and initialize result just after the GC.